numpy/core/_add_newdocs.py

```diff
@@ -1041,7 +1041,8 @@
         elements is also ignored.
 
         .. deprecated:: 1.14
-            If this argument is not provided, `fromstring` falls back on the
+            Passing this argument as an empty string is deprecated.
+            If `seq == ''`, `fromstring` falls back on the
             behaviour of `frombuffer` after encoding unicode string inputs as
             either utf-8 (python 3), or the default encoding (python 2).
```

Comment:
We are now python3 only. I find the phrasing here still a bit cryptic. Maybe something like 
> `sep` is now required and must be non-empty since passing `sep=''` will trigger the binary mode of `PyArray_FromString`. This encodes `str` as utf-8 and passes it to `frombuffer`". If this is your intent, do so explicitly with ``frombuffer(str.encode('utf-8'), dtype, count)``

xref issue gh-9484.

Reply:
I do wonder whether we want to remove python 2 references from the docs. The reality is that python 2 users will continue to click on the first google result for numpy documentation, and probably not pay too much attention to the version number. Until we can get the version switcher into the docs webpages, cutting out python 2 references strikes me as doing more harm than good.

Reply:
OK, how about 

> `sep` is now required and must be non-empty since passing `sep=''` will trigger the binary mode of `PyArray_FromString`. This encodes `str` and passes it to `frombuffer`". If this is your intent, do so explicitly with ``frombuffer(str.encode('utf-8'), dtype, count)`` (or ``frombuffer(str.encode(), dtype, count)`` on python2)

Reply:
"required and must be" seems like the description we'd apply after we finish deprecating, not the one to use while the deprecation is in progress. Perhaps _"Passing `sep=''` will trigger the deprecated binary mode of this function, which encodes `str` and passes it to `frombuffer`. If this is your intent ..."_

Reply:
Really, there are two parallel deprecations going on here:

* `np.fromstring(some_bytes_obj)` is deprecated because it's a misleading way to spell `np.frombytes(some_bytes_obj)`
* `np.fromstring(some_unicode_obj)` is deprecated because `np.frombytes(some_unicode_obj.encode('utf8'))` is nonsense

The currently wording covers both these cases, but is perhaps non-obvious about doing so.

Your wording, including the suggestion of replacing with `frombuffer(str.encode('utf-8'), dtype, count)` is inaccurate, the actual alternative is `frombuffer(s.encode('utf-8') if not isinstance(s, bytes) else s, dtype, count)`.

Unfortunately, often the behavior being deprecated is such a weird thing to do that even describing it is hard.